modules/images/webp-uploads/load.php

```diff
@@ -285,3 +284,53 @@ function webp_uploads_remove_sources_files( $attachment_id ) {
 }
 
 add_action( 'delete_attachment', 'webp_uploads_remove_sources_files', 10, 1 );
+
+/**
+ * Filter on `wp_get_missing_image_subsizes` acting as an action for the logic of the plugin
+ * to determine if additional mime types still need to be created.
+ *
+ * @since n.e.x.t
+ *
+ * @see   wp_get_missing_image_subsizes()
+ *
+ * @param array[] $missing_sizes Associative array of arrays of image sub-size.
+ * @param array   $image_meta    The metadata from the image.
+ * @param int     $attachment_id The ID of the attachment.
+ *
+ * @return array[] $missing_sizes Associative array of arrays of image sub-size.
+ */
+function webp_uploads_wp_get_missing_image_subsizes( $missing_sizes, $image_meta, $attachment_id ) {
+	if ( _webp_uploads_is_valid_ajax_for_image_sizes() || _webp_uploads_is_valid_rest_for_post_process( file_get_contents( 'php://input' ) ) ) {
+		webp_uploads_create_sources_property( $image_meta, $attachment_id );
+	}
+
+	return $missing_sizes;
+}
+
+/**
+ * Determine if the current request is a valid request to process missing image sizes, executed
+ * via ajax.
+ *
+ * @return bool `true` if the current request is a valid ajax request to recreate missing image sizes.
+ */
+function _webp_uploads_is_valid_ajax_for_image_sizes() {
+	return wp_doing_ajax() && isset( $_REQUEST['action'], $_REQUEST['attachment_id'] ) && 'media-create-image-subsizes' === $_REQUEST['action'] && $_REQUEST['attachment_id'] > 0;
+}
+
+/**
+ * Determine if the provided REST request goes to the appropriate endpoint with the right action.
+ *
+ * @param string $body The body from the request.
+ *
+ * @return bool `true` if the current request is a valid REST request to process
+ */
+function _webp_uploads_is_valid_rest_for_post_process( $body = '' ) {
+	$matches     = array();
+	$body        = json_decode( $body, true );
+	$valid_rest  = defined( 'REST_REQUEST' ) && REST_REQUEST;
+	$valid_route = $valid_rest && isset( $GLOBALS['wp'], $GLOBALS['wp']->query_vars['rest_route'] ) && preg_match( '/media\/\d+\/post-process/', $GLOBALS['wp']->query_vars['rest_route'], $matches );
+
+	return $valid_route && ! empty( $matches ) && ! empty( $body['action'] ) && 'create-image-subsizes' === $body['action'];
+}
```

Comment:
Based on what we discussed yesterday, there is a cleaner solution that having these two functions - we shouldn't need to specifically check whether it's an AJAX or REST request. What about if any other logic generates image sub-sizes via `wp_update_image_subsizes()`? Your current code wouldn't support that.

Let's remove these functions and instead implement a more stable solution based on what we talked about yesterday:
* In `webp_uploads_create_sources_property()`, after the initial checks but before starting the `foreach` loop, we can set a custom meta flag to indicate that this attachment is awaiting additional MIME type generation. We could use e.g. `_webp_uploads_awaiting_additional_mime_subsizes` as name.
* We should set this meta key on the attachment (e.g. via `update_post_meta()`), only if it is not set yet, so that it is not set over and over again on subsequent calls.
* Right after the foreach loop (i.e. if the whole process was successful), we should unset that flag again (e.g. via `delete_post_meta()`).
* In your above `webp_uploads_wp_get_missing_image_subsizes()` function, we can then check if `empty( $missing_sizes ) && get_post_meta( $attachment_id, '_webp_uploads_awaiting_additional_mime_subsizes', true )` instead of the less reliable check for whether this is an AJAX or REST request.

Reply:
Actually, I was looking at the code, and having the meta just does not provide much value add is a meta that won't be used after the merge in core so there's no real benefit in keeping it together as if the function is executed and all sizes are looped the image sizes would be updated only if they do not exist which is exactly what we aim with the meta but without the need to define a new meta key and keep track of it in multiple places.

On the other hand, we know from the code the retry mechanism would be executed only from 2 places from ajax and REST, so the conditionals are in place to prevent exactly that execute the logic to create missing images in a context where was not requested, having the conditionals ensure that even if this code is executed from outside of those 2 contexts we are not executing that logic.

However, if the aim is to review this logic anytime this function is called regardless of the context we can just remove the conditionals and execute the function `webp_uploads_create_sources_property`as there's no real need to have an additional meta if the logic requires image generation would do so if not it would just loop into every image size skipping over every single one of them.

Reply:
@mitogh 
> is a meta that won't be used after the merge in core so there's no real benefit

Indeed this is a workaround which we only need because there's no appropriate action for this in core. But the fact that we won't need it later in core IMO isn't a good argument against it. The current logic you have here checking for AJAX or REST request won't be needed in WP core either.

Also, the meta key will always be set on an attachment only temporarily and be removed as soon as all the image formats generation has completed, so it won't even stay around in the future, so we're not "polluting" the database in any way.

> On the other hand, we know from the code the retry mechanism would be executed only from 2 places from ajax and REST

I think this is a problematic conclusion: We know that that is the case from WP core itself, but we have to keep in mind the existence of plugins. Any plugin may integrate with this functionality, a plugin could potentially call `wp_update_image_subsizes()` directly. For example, what if there was a GraphQL plugin facilitating image uploads and using this function? The only way to have the implementation here reliable is to find a solution that works only based on the `wp_update_image_subsizes()` function, we can't really make an assumption where this is called or not.

> we can just remove the conditionals and execute the function `webp_uploads_create_sources_property`as there's no real need to have an additional meta

That won't work, because then the code may run unexpectedly and accidentally create image sizes even when the `wp_get_missing_image_subsizes()` function is called _just_ to list the missing sub-sizes (which happens e.g. in one of the REST endpoints). Setting an explicit flag (e.g. using the meta key) in our logic to generate the additional format sub-sizes, at a point when we _really_ know it should generate them, makes this happen on-demand.

Without such a flag, even when somebody calls the REST API to get information about the image 3 months after the upload, our logic would still try to run. Yes, it wouldn't actually generate image sizes then since they will all exist already, but even attempting to run the logic in this context is wasteful - and potentially even problematic from a security perspective, since that would mean a write action is happening during a REST API `GET` (read) request.

Reply:
@mitogh Actually, I just tried another solution using `debug_backtrace()` that should mitigate both your and my concerns above. Really, what we need to check is whether `wp_get_missing_image_subsizes()` was called within `wp_update_image_subsizes()` - that's our condition. It's still hacky, but that's really the most reliable solution I think. Here's some demo code I used to try this out:
```php
add_filter(
	'wp_get_missing_image_subsizes',
	function( $missing_sizes, $image_meta, $attachment_id ) {
		if ( ! empty( $missing_sizes ) ) {
			return $missing_sizes;
		}
		// Should be at position 4, but look into the first 10 just to be safe.
		$trace = debug_backtrace( DEBUG_BACKTRACE_IGNORE_ARGS, 10 );
		foreach ( $trace as $element ) {
			if ( isset( $element['function'] ) && 'wp_update_image_subsizes' === $element['function'] ) {
				// Call our function to generate more image formats here.
				break;
			}
		}
		return $missing_sizes;
	},
	10,
	3
);
```

It's a bit weird, but I actually think this is less hacky than the two other solutions we've looked at so far. What do you think?

Reply:
> I think this is a problematic conclusion: We know that that is the case from WP core itself, but we have to keep in mind the existence of plugins. Any plugin may integrate with this functionality, a plugin could potentially call wp_update_image_subsizes() directly. For example, what if there was a GraphQL plugin facilitating image uploads and using this function? The only way to have the implementation here reliable is to find a solution that works only based on the wp_update_image_subsizes() function, we can't really make an assumption where this is called or not.

Correct however we are under the assumption that the retry mechanism would be executed from within the core mechanism, if an upload is processed via a GraphQL plugin and the request fails there's no retry mechanism there that would be used so is a safer assumption to introduce the retry logic only in 2 places where we know for a fact is happening Ajax and REST

From my perspective, the way it should work is if a plugin fails to upload it should rely on WP core to add the missing images using the same APIs from WP Core, not the other way around fixing the plugins from the WP Core side. In this case, a plugin failing to process images should use the REST endpoints created for this reason, Which then would trigger the logic that we have in place.

> Without such a flag, even when somebody calls the REST API to get information about the image 3 months after the upload, our logic would still try to run. Yes, it wouldn't actually generate image sizes then since they will all exist already, but even attempting to run the logic in this context is wasteful - and potentially even problematic from a security perspective, since that would mean a write action is happening during a REST API GET (read) request.

Right, however even with a meta key in place this also can happen as if a meta exists and the maximum number of retries are executed the meta won't be removed, next time this function is executed from a GET context the meta would execute the logic to regenerate the images even when that's not the expected behavior.

From my perspective instead of limiting the scope from where this can be executed, we allow for safer usage of the function in other contexts like listing all missing sizes from an image and not actually triggering the regeneration of image sizes.

Reply:
Fantastic, the `debug_backtrace`  seems like the way to go :100: agree on that one.